internal/controller/nats/deprovisioner.go

```diff
@@ -2,33 +2,80 @@ package nats
 
 import (
 	"context"
+	"fmt"
 
 	natsv1alpha1 "github.com/kyma-project/nats-manager/api/v1alpha1"
 	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
+const (
+	StreamExistsErrorMsg = "Cannot delete NATS cluster as stream exists"
+	natsClientPort       = 4222
+	InstanceLabelKey     = "app.kubernetes.io/instance"
+)
+
 func (r *Reconciler) handleNATSDeletion(ctx context.Context, nats *natsv1alpha1.NATS,
 	log *zap.SugaredLogger) (ctrl.Result, error) {
 	// skip reconciliation for deletion if the finalizer is not set.
 	if !r.containsFinalizer(nats) {
-		log.Debugf("skipped reconciliation for deletion as finalize is not set.")
+		log.Debugf("skipped reconciliation for deletion as finalizer is not set.")
 		return ctrl.Result{}, nil
 	}
 
 	r.logger.Info("Deleting the NATS")
 	nats.Status.SetStateDeleting()
 
-	// get NATS resources to de-provision
-	instance, err := r.initNATSInstance(ctx, nats, log)
+	// create a new NATS client instance
+	if err := r.createAndConnectNatsClient(nats); err != nil {
+		// delete a PVC if NATS client cannot be created
+		return r.deletePVCsAndRemoveFinalizer(ctx, nats, r.logger)
+	}
+	// check if NATS JetStream stream exists
+	streamExists, err := r.getNatsClient(nats).StreamExists()
 	if err != nil {
-		return ctrl.Result{}, err
+		// delete a PVC if NATS client cannot be created
+		return r.deletePVCsAndRemoveFinalizer(ctx, nats, r.logger)
 	}
+	if streamExists {
+		// if a stream exists, do not delete the NATS cluster
+		nats.Status.UpdateConditionDeletion(metav1.ConditionFalse,
+			natsv1alpha1.ConditionReasonDeletionError, StreamExistsErrorMsg)
+		return ctrl.Result{Requeue: true}, r.syncNATSStatus(ctx, nats, log)
+	}
+
+	return r.deletePVCsAndRemoveFinalizer(ctx, nats, r.logger)
+}
 
-	// delete all NATS resources
-	if err = r.natsManager.DeleteInstance(ctx, instance); err != nil {
-		return ctrl.Result{}, r.syncNATSStatusWithErr(ctx, nats, err, log)
+// create a new NATS client instance and connect to the NATS server.
+func (r *Reconciler) createAndConnectNatsClient(nats *natsv1alpha1.NATS) error {
+	// create a new instance if it does not exist
+	if r.getNatsClient(nats) == nil {
+		r.setNatsClient(nats, NewNatsClient(&natsConfig{
+			URL: fmt.Sprintf("nats://%s.%s.svc.cluster.local:%d", nats.Name, nats.Namespace, natsClientPort),
+		}))
 	}
+	return r.getNatsClient(nats).Init()
+}
 
+func (r *Reconciler) deletePVCsAndRemoveFinalizer(ctx context.Context,
+	nats *natsv1alpha1.NATS, log *zap.SugaredLogger) (ctrl.Result, error) {
```

Comment:
Does this method need to reside in controller pkg? or would it be a good idea to keep all NATS related methods in a separate `pkg/nats` package?